geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/geotools/Conversions.scala

```diff
@@ -94,13 +100,78 @@ object Conversions {
     }
   }
 
-  import scala.collection.JavaConversions._
-  implicit class RichAttributeDescriptor(val attr: AttributeDescriptor) extends AnyVal {
-    def isIndexed = attr.getUserData.getOrElse("index", false).asInstanceOf[java.lang.Boolean]
-    def isCollection = classOf[JCollection[_]].isAssignableFrom(attr.getType.getBinding)
-    def isMap = classOf[JMap[_, _]].isAssignableFrom(attr.getType.getBinding)
+}
+
+/**
+ * Contains GeoMesa specific attribute descriptor information
+ */
+object RichAttributeDescriptors {
+
+  import org.locationtech.geomesa.utils.geotools.SimpleFeatureTypes._
+
+  implicit class RichAttributeDescriptor(val ad: AttributeDescriptor) extends AnyVal {
+
+    def setIndexCoverage(coverage: IndexCoverage): Unit = ad.getUserData.put(OPT_INDEX, coverage.toString)
+
+    def getIndexCoverage(): IndexCoverage =
+      Option(ad.getUserData.get(OPT_INDEX).asInstanceOf[String])
+          .flatMap(c => Try(IndexCoverage.withName(c)).toOption).getOrElse(IndexCoverage.NONE)
+
+    def setIndexValue(indexValue: Boolean): Unit =
+      ad.getUserData.put(OPT_INDEX_VALUE, new java.lang.Boolean(indexValue))
+
+    def getIndexValue(): Boolean =
+      Option(ad.getUserData.get(OPT_INDEX_VALUE).asInstanceOf[Boolean]).getOrElse(false)
+
+    def setCardinality(cardinality: Cardinality): Unit =
+      ad.getUserData.put(OPT_CARDINALITY, cardinality.toString)
+
+    def getCardinality(): Cardinality =
+      Option(ad.getUserData.get(OPT_CARDINALITY).asInstanceOf[String])
+          .flatMap(c => Try(Cardinality.withName(c)).toOption).getOrElse(Cardinality.UNKNOWN)
+
+    def setCollectionType(typ: Class[_]): Unit = ad.getUserData.put(USER_DATA_LIST_TYPE, typ)
+
+    def getCollectionType(): Option[Class[_]] =
+      Option(ad.getUserData.get(USER_DATA_LIST_TYPE)).map(_.asInstanceOf[Class[_]])
+
+    def setMapTypes(keyType: Class[_], valueType: Class[_]): Unit = {
+      ad.getUserData.put(USER_DATA_MAP_KEY_TYPE, keyType)
+      ad.getUserData.put(USER_DATA_MAP_VALUE_TYPE, valueType)
+    }
+
+    def getMapTypes(): Option[(Class[_], Class[_])] = for {
+      keyClass   <- Option(ad.getUserData.get(USER_DATA_MAP_KEY_TYPE))
+      valueClass <- Option(ad.getUserData.get(USER_DATA_MAP_VALUE_TYPE))
+    } yield {
+      (keyClass.asInstanceOf[Class[_]], valueClass.asInstanceOf[Class[_]])
+    }
+
+    def isIndexed = getIndexCoverage() match {
+      case IndexCoverage.FULL | IndexCoverage.JOIN => true
```

Comment:
yeah, you just can't bind to a variable if you use it:
http://stackoverflow.com/questions/1837754/match-multiple-cases-classes-in-scala